rectools/models/nn/transformers/similarity.py

```diff
@@ -0,0 +1,110 @@
+import typing as tp
+
+import numpy as np
+import torch
+import torch.nn as nn
+from scipy import sparse
+
+from rectools.models.base import InternalRecoTriplet
+from rectools.models.rank import Distance, TorchRanker
+from rectools.types import InternalIdsArray
+
+
+class SimilarityModuleBase(nn.Module):
+    """Similarity module base."""
+
+    def _get_full_catalog_logits(self, session_embs: torch.Tensor, item_embs: torch.Tensor) -> torch.Tensor:
+        raise NotImplementedError()
+
+    def _get_pos_neg_logits(
+        self, session_embs: torch.Tensor, item_embs: torch.Tensor, item_ids: torch.Tensor
+    ) -> torch.Tensor:
+        raise NotImplementedError()
+
+    def forward(
+        self, session_embs: torch.Tensor, item_embs: torch.Tensor, item_ids: tp.Optional[torch.Tensor] = None
+    ) -> torch.Tensor:
+        """Forward pass to get logits."""
+        raise NotImplementedError()
+
+    def _recommend_u2i(
+        self,
+        user_embs: torch.Tensor,
+        item_embs: torch.Tensor,
+        user_ids: InternalIdsArray,
+        k: int,
+        sorted_item_ids_to_recommend: InternalIdsArray,
+        ui_csr_for_filter: tp.Optional[sparse.csr_matrix],
+    ) -> InternalRecoTriplet:
+        """Recommend to users."""
+        raise NotImplementedError()
+
+
+class SimilarityModuleDistance(SimilarityModuleBase):
+    """Similarity module distance."""
+
+    dist_available: tp.List[str] = ["dot", "cosine"]
+    # dist_available_values: tp.List[int] = [Distance.DOT.value, Distance.COSINE.value]
+    epsilon_cosine_dist: float = 1e-8
+
+    def __init__(self, dist: str = "dot") -> None:
+        super().__init__()
+        if dist not in self.dist_available:
+            raise ValueError("`dist` can only be either `dot` or `cosine`.")
+
+        self.dist = dist
```

Comment:
```suggestion
        self.distance = Distance(distance)
```
Should work? Then we don't need an `if` in `_recommend_u2i`

Reply:
Doesn't work
`TypeError: __call__() missing 1 required positional argument: 'value'`

It can be work as `Distance(1)`.

Reply:
Let's update there:
```
class Distance(str, Enum):
    """Distance metric"""

    DOT = "dot"  # Bigger value means closer vectors
    COSINE = "cosine"  # Bigger value means closer vectors
    EUCLIDEAN = "euclidean"  # Smaller value means closer vectors
```